In [1]:
import pandas as pd
%matplotlib inline

In [2]:
df_rule = pd.read_csv('./data/reaction_rule.csv',index_col=0)

In [3]:
df_rule.shape

(1435, 7053)

In [4]:
df_rule.head()

R08505  R08852  R08851  R07882  R07885  R08856  R08503  R08854  \
BO            0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
Bc(c)c        0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
BrC(Br)Br     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
C             0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
C#C           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

           R08859  R08516   ...    R08253  R08678  R08673  R07152  R08671  \
BO            0.0     0.0   ...       0.0     0.0     0.0     0.0     0.0   
Bc(c)c        0.0     0.0   ...       0.0     0.0     0.0     0.0     0.0   
BrC(Br)Br     0.0     0.0   ...       0.0     0.0     0.0     0.0     0.0   
C             0.0     0.0   ...       0.0     0.0     0.0     0.0     0.0   
C#C           0.0     0.0   ...       0.0     0.0     0.0     0.0     0.0   

           R08670  R08677  R08676  R08675  R08674  
BO            0.0     0.0     0.0     0.0     0.0  
Bc(c)c        0.0     0.0     0.0     0.0     0.0  
BrC(Br)Br     0.0     0.0     0.0     0.0     0.0  
C             0.0     0.0     0.0     0.0     0.0  
C#C           0.0     0.0     0.0     0.0     0.0  

[5 rows x 7053 columns]

In [5]:
# read groups that exist in thermodynamic data
df_groups = pd.read_csv('./data/group_names_TECRDB.txt',header=None)

In [6]:
groups = df_groups[0].tolist()

In [7]:
len(groups)

263

# find reactions that do not decompose into groups that are not defined by TECRDB

In [8]:
df_rule_2 = df_rule.drop(groups)

In [9]:
df_rule_2.shape

(1172, 7053)

In [10]:
df_rule_2.head()

R08505  R08852  R08851  R07882  R07885  R08856  R08503  R08854  \
BO            0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
Bc(c)c        0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
BrC(Br)Br     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
C#C           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
C#CBr         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

           R08859  R08516   ...    R08253  R08678  R08673  R07152  R08671  \
BO            0.0     0.0   ...       0.0     0.0     0.0     0.0     0.0   
Bc(c)c        0.0     0.0   ...       0.0     0.0     0.0     0.0     0.0   
BrC(Br)Br     0.0     0.0   ...       0.0     0.0     0.0     0.0     0.0   
C#C           0.0     0.0   ...       0.0     0.0     0.0     0.0     0.0   
C#CBr         0.0     0.0   ...       0.0     0.0     0.0     0.0     0.0   

           R08670  R08677  R08676  R08675  R08674  
BO            0.0     0.0     0.0     0.0     0.0  
Bc(c)c        0.0     0.0     0.0     0.0     0.0  
BrC(Br)Br     0.0     0.0     0.0     0.0     0.0  
C#C           0.0     0.0     0.0     0.0     0.0  
C#CBr         0.0     0.0     0.0     0.0     0.0  

[5 rows x 7053 columns]

- Find reactions that have groups outside of TECRDB (df_rule_3)
- All - df_rule_3

In [11]:
df_rule_3 = df_rule_2.loc[:, (df_rule_2 != 0).any(axis=0)]

In [12]:
df_rule_3.shape

(1172, 2166)

In [13]:
result = 7053 - 2166

In [14]:
result

4887

# An example of the reaction that cannot be describle by the 263 groups

In [15]:
df_rule_3.head()

R08859  R03359  R03358  R01267  R01269  R02006  R02007  R02004  \
BO            0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
Bc(c)c        0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
BrC(Br)Br     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
C#C           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
C#CBr         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

           R02003  R02008   ...    R08250  R08672  R08253  R08678  R08673  \
BO            0.0     0.0   ...       0.0     0.0     0.0     0.0     0.0   
Bc(c)c        0.0     0.0   ...       0.0     0.0     0.0     0.0     0.0   
BrC(Br)Br     0.0     0.0   ...       0.0     0.0     0.0     0.0     0.0   
C#C           0.0     0.0   ...       0.0     0.0     0.0     0.0     0.0   
C#CBr         0.0     0.0   ...       0.0     0.0     0.0     0.0     0.0   

           R08671  R08670  R08677  R08675  R08674  
BO            0.0     0.0     0.0     0.0     0.0  
Bc(c)c        0.0     0.0     0.0     0.0     0.0  
BrC(Br)Br     0.0     0.0     0.0     0.0     0.0  
C#C           0.0     0.0     0.0     0.0     0.0  
C#CBr         0.0     0.0     0.0     0.0     0.0  

[5 rows x 2166 columns]

In [21]:
example_df = df_rule_3['R08250'].to_frame()
example_df[~(example_df == 0).any(axis=1)]

R08250
CNN     1.0

In [22]:
example_df_1 = df_rule['R08250'].to_frame()
example_df_1[~(example_df_1 == 0).any(axis=1)]

R08250
CC(=O)S    -1.0
CC(N)=O     1.0
CNN         1.0
CS          1.0
CSC        -1.0
NN         -1.0

# Example of no group cahnge

In [23]:
example_df_2 = df_rule['R00889'].to_frame()
example_df_2[~(example_df_2 == 0).any(axis=1)]

R00889
C[C@@H](C)O    -2.0
C[C@H](C)O      2.0